### PR TITLE
[SPARK-9460] Fix prefix generation for UTF8String.

### PR DESCRIPTION
**Pull request by rxin**
Previously we could be getting garbage data if the number of bytes is 0, or on JVMs that are 4 byte aligned, or when compressedoops is on.

### CODE REVIEW

**Comment by rxin** [comment]
cc @davies and @tellison

**Comment by rednaxelafx** [comment]
Yes, this change looks good to me, for a "semantically safe bet" on the object layout.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -141,7 +141,19 @@ public int numChars() {
...
+    // If size is greater than 4, assume we have at least 8 bytes of data to fetch.
+    long p;
+    if (numBytes > 4) {
+      p = PlatformDependent.UNSAFE.getLong(base, offset);
+    } else if (numBytes > 0) {
+      p = (long) PlatformDependent.UNSAFE.getInt(base, offset);
+    } else {
+      p = 0;
+    }
     p = java.lang.Long.reverseBytes(p);
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Should we check the endian-ness? I think UNSAFE.getLong use the native endian.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -141,7 +154,24 @@ public int numChars() {
...
   public long getPrefix() {
-    long p = PlatformDependent.UNSAFE.getLong(base, offset);
+    // Since JVMs are either 4-byte aligned or 8-byte aligned, we check the size of the string.
+    // If size is 0, just return 0.
+    // If size is between 0 and 4 (inclusive), assume data is 4-byte aligned under the hood and
+    // use a getInt to fetch the prefix.
+    // If size is greater than 4, assume we have at least 8 bytes of data to fetch.
+    // After getting the data, we use a mask to mask out data that is not part of the string.
+    long p;
+    if (numBytes >= 8) {
```
Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: this currently won't work for big endian. can somebody from ibm submit a patch to make it work? I don't have the environment to test.

Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Maybe we can just merge this one, add a comment and create a JIRA for big-endian.

Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Yup that's a good idea.

Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Here's the ticket https://issues.apache.org/jira/browse/SPARK-9483

**Comment by rxin** [comment]
Merging this since the failed test is unrelated.

**Pull request closed without merging**